### ISSUE TITLE
talos_cluster_health times out when adding new nodes

### ISSUE THREAD

**Issue by rmvangun** [siderolabs/terraform-provider-talos#221]
When creating a basic Talos cluster with Terraform, I think there's a kind of logical problem that blocks health checks in the plan phase when adding new nodes. The `talos_cluster_health` data source takes a list of all controlplane and workers. The first run, this is fine, the nodes are configured, and the health check passes. However, the second time, let's say you increase your worker node count. Now, the health check times out _in the plan phase_. The reason is that it now expects an additional worker node to be available. However, that node hasn't yet been given machine configuration in the apply step.

So, the health check runs during the plan phase, expects a new healthy worker node that has yet to be configured, and so fails the check.

Here's the cluster health config:

```
data "talos_cluster_health" "this" {
  depends_on = [talos_cluster_kubeconfig.this]

  client_configuration   = talos_machine_secrets.this.client_configuration
  control_plane_nodes    = module.controlplanes.*.node
  worker_nodes           = module.workers.*.node
  endpoints              = module.controlplanes.*.endpoint
}
```

Add a new node, and the plan times out with: `data.talos_cluster_health.this: Still reading... [20s elapsed]`

Not sure how to fix this, wish the health check just knew not to check on nodes that have yet to be configured.

**Comment by nebula-it** [comment]
Facing the exact same issue here. We need health check only on the first run to confirm cluster is up before deploying CNI. There should be `skip_node_check` that can be used so it doesnt worry anything about nodes, just makes sure cluster endpoint is healthy.

**Comment by frezbo** [comment]
you'd need to use https://registry.terraform.io/providers/siderolabs/talos/latest/docs/data-sources/cluster_health#skip_kubernetes_checks-1

**Comment by nebula-it** [comment]
That does not work, it keeps trying to connect to the IP of new worker node (Since we are passing IPs of all workers nodes(calculated in a local var) the IP of new worker node is in there) and then times out

**Comment by jonkerj** [comment]
> you'd need to use https://registry.terraform.io/providers/siderolabs/talos/latest/docs/data-sources/cluster_health#skip_kubernetes_checks-1

That does not seem to work in my case: the cluster is healthy Talos-wise, but nodes are not ready due to the CNI not being installed yet. For some reason, health is timing out (and I'm using `skip_kubernetes_checks = true`) indefinately.

I'd like to have the Cilium helm-release (that is going to fix that) depend on talos' health.

**Comment by nebula-it** [comment]
Hey @jonkerj, It does work for us without any CNI in place. We are also using it as pre-check to do the helm install. To confirm are you passing the IP address of all CP and worker nodes to it? Whats the error message you get at the end, when it fails/timesout?

**Comment by nebula-it** [comment]
@frezbo Can we please get an update on this? As of now we have to comment out health data if we need to add new worker nodes, which makes for a really cumbersome workflow.

**Comment by FlippingBinary** [comment]
Is anyone running into this problem using `talos_cluster_health` for a reason _other_ than to wait on bootstrapping to finish after using `talos_machine_bootstrap`?

Now, I'm admittedly not an expert in Terraform or Kubernetes/etcd, but my understanding is that Terraform's data sources are there to gather the data that is necessary to build a plan of action ahead of time, not to control the flow during its implementation. If `talos_machine_bootstrap` might return without error even though the machine failed to bootstrap (i.e. it's not part of the cluster after it finishes rebooting), perhaps that's a problem. It should be waiting until it confirms the machine actually bootstrapped the cluster before returning.

Having said that, I wonder if perhaps the resources in this provider in their current state are mapping too closely to the Talos CLI. Terraform is a declarative tool, so it works best when it is used to define the desired state of resources, not the actions that should be taken on them.

Ideally, there would be a `talos_cluster` resource and a `talos_node` resource. The `talos_cluster` resource would be declared with a property that lists the controller nodes by a readonly identifier of the corresponding `talos_node` resources. Same with the worker nodes. This would create a natural dependency that Terraform understands so that it would never try to bootstrap the cluster before configuring all of its nodes.

Like this:

```hcl
resource "talos_cluster" "this" {
  controlplane_nodes = [ talos_node.controller1.id ]
  worker_nodes = [ talos_node.worker1.id, talos_node.worker2.id ]
  # ...
}

resource "talos_node" "controller1" {
  # ...
}

resource "talos_node" "worker1" {
  # ...
}

resource "talos_node" "worker2" {
  # ...
}
```

Under the hood, this Terraform provider would need to do whatever is necessary to alter the configuration of the cluster and its nodes to match the new state defined in the Terraform config regardless of its previous state. That might include bootstrapping, applying configuration changes, etc., but it's up to the plugin to figure that out (with the benefit of Terraform's record of previous state), not the configuration. When it finishes creating one resource, any other resource that depends on it should be able to rely upon the fact that the infrastructure it represents is healthy.

As another temporary workaround, Terraform's `-target` command line argument can be used to select the `talos_machine_bootstrap` resource (and/or any others) that need to be created before `talos_cluster_health` will complete successfully. That allows a partial application of the configuration. The full configuration can be applied after the partial one completes. It's not ideal, but I'm finding it easier to do that than commenting out sections and remembering to put them back.
